modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java

```diff
@@ -315,7 +316,7 @@ public RelDataType createCustomType(Type type) {
     public RelDataType createCustomType(Type type, boolean nullable) {
         if (UUID.class == type)
             return canonize(new UuidType(nullable));
-        else if (Object.class == type)
+        else if (Object.class == type || (type instanceof Class && BinaryObject.class.isAssignableFrom((Class<?>)type)))
```

Comment:
`type instanceof Class` looks useless

Reply:
Why? In case of not `Class` type we can get an exception on cast to `Class`, so it's safer to check type.

Reply:
Is `BinaryObject.class.isAssignableFrom` not enough? Do we have any tests for that?

Reply:
I think we can't get any problem here. Calcite expects a class in `createType(Type type)`:
`if (!(type instanceof Class)) {
      throw new UnsupportedOperationException("TODO: implement " + type);
    }
    final Class clazz = (Class) type;`

An `assert` looks enough to me. Up to you,

Reply:
In `createType` we call `super.createType` if custom type is null. So `type` other than `Class` can be processed by superclass.